near-sdk-macros/src/core_impl/info_extractor/impl_item_method_info.rs

```diff
@@ -1,22 +1,48 @@
 use crate::core_impl::info_extractor::AttrSigInfo;
-use syn::{ImplItemMethod, Type, Visibility};
+use syn::{spanned::Spanned, Error, ImplItemMethod, Signature, Type, Visibility};
 
 /// Information extracted from `ImplItemMethod`.
 pub struct ImplItemMethodInfo {
     /// Information on the attributes and the signature of the method.
     pub attr_signature_info: AttrSigInfo,
-    /// Whether method has `pub` modifier.
-    pub is_public: bool,
     /// The type of the contract struct.
     pub struct_type: Type,
 }
 
 impl ImplItemMethodInfo {
+    fn check_sig_modifiers(sig: &Signature) -> syn::Result<()> {
+        if sig.asyncness.is_some() {
+            return Err(Error::new(sig.span(), "Contract API is not allowed to be async."));
+        }
+        if sig.abi.is_some() {
+            return Err(Error::new(
+                sig.span(),
+                "Contract API is not allowed to have binary interface.",
+            ));
+        }
+        if sig.variadic.is_some() {
+            return Err(Error::new(
+                sig.span(),
+                "Contract API is not allowed to have variadic arguments.",
+            ));
+        }
+
+        Ok(())
+    }
```

Comment:
Aren't these checks useful for `TraitItemMethodInfo`?

Reply:
Yes, good catch. This also removes the side effect of not checking `#[ext_contract]` which you mentioned (was not intended). I think it makes more sense to prohibit that stuff until we find an actual use for them.